### DISCUSSION TITLE
Using Spot instances in AWS Fargate with Artillery

### DISCUSSION THREAD

**Discussion by moufmouf** [artilleryio/artillery#2252] [Artillery]
Hi team and first of all, thanks a lot for Artillery. I've been using it 3 years ago. I'm coming back today and woah! It's incredible how it evolved.

Now to my question. I'm using the AWS ECS Fargate cluster to run tests on Playwright. Everything seems to be running fine so far, and it was really easy to set up.

I'm wondering if there is a way to tell Playwright to use Spot instances by default (as described here: https://aws.amazon.com/fr/ec2/spot/containers-for-less/get-started/)
From what I see in the logs, Artillery uses standard instances (I may be wrong though). Is it possible to use Spot instances instead? (they are cheap, they can potentially fail but that's not a big deal for a performance test).
If not, do you think it is complex to implement? (I'm willing to help write a PR if it sounds possible)

Thanks!

**Comment by hassy** [comment]
Thanks @moufmouf, that's really nice to hear! :) 

I *think* it should just work. If you have an ECS cluster using Spot Instances, you should be able to tell Artillery to use it with the [`--cluster`](https://www.artillery.io/docs/reference/cli/run-fargate#fargate-specific-flags) flag. Artillery itself uses standard AWS ECS APIs to schedule tasks on ECS/Fargate so it shouldn't matter whether it's a reserved or spot cluster, as long as there's enough capacity to schedule the tasks for the test run.

Let us know if that doesn't work though and we can investigate

**Reply by bernardobridge** [reply]
Hey @moufmouf , just adding some additional context here. There's a chance this won't actually work, due to this sentence in the [AWS SDK docs](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/ECS.html#runTask-property):

>If a capacityProviderStrategy is specified, the launchType parameter must be omitted. If no capacityProviderStrategy or launchType is specified, the defaultCapacityProviderStrategy for the cluster is used.

As we set launchType to `FARGATE`, it might not be able to use the default capacity provider strategy. 

Let us know if that's the case, and we'll discuss a way to make this work! 🙇‍♂️

**Reply by moufmouf** [reply]
Hey @hassy , hey @bernardobridge ,

Thanks a lot for all your advices!

I've tested creating a new ECS cluster. In this cluster, I configured the default capacity provider to FARGATE_SPOT, as shown in the picture below.

![image](https://github.com/artilleryio/artillery/assets/1290952/e93f2de9-1531-4fcc-8dfa-199a97c99870)

Now, I'm launching the test with the "--cluster" parameter.
Tasks are correctly starting.

When I look at the task details, I see this:

![image](https://github.com/artilleryio/artillery/assets/1290952/c65d06b4-8d6d-4c02-82d8-8837b45e1869)

Launch type is "FARGATE". This is ok for me, as according to [the documentation here](https://aws.amazon.com/fr/blogs/compute/deep-dive-into-fargate-spot-to-run-your-ecs-tasks-for-up-to-70-less/), the launch type is "FARGATE" even for spot instances.
However, in the "Capacity provider" zone, I did expect to see "FARGATE SPOT", and there is nothing displayed.

So I must admit I have no idea if the task is running as a SPOT instance or not :shrug: 

Any idea? Or can you point me to the code that starts the tasks?

**Reply by bernardobridge** [reply]
Hey @moufmouf ! I don't think it's using spot instances, as it would show `FARGATE_SPOT` under capacity provider if it did, I believe. I think what I [described here](https://github.com/artilleryio/artillery/discussions/2252#discussioncomment-7390715) is likely the reason behind this - we set `FARGATE` as the launch type, so it doesn't use the default capacity provider. 

The implementation code for Fargate is currently in a separate private repo (due to it formerly being `artillery pro`), but we're working on changing that soon. In the meantime, we'll discuss how to make support for `FARGATE_SPOT` instances work, and implement this when we can.

The regular Fargate instances are still quite cheap though (unless you are running them for very very long periods with very large `--count`), so hopefully you'll still be able to run tests while we don't provide this 😄. Hope that helps 👍

**Reply by moufmouf** [reply]
Thanks a lot @bernardobridge !

**Comment by mrister** [comment]
Any update on this. We would like to use it for load testing meaning using many containers... so FARGATE_SPOT is a good choice. How to do this ?

**Reply by hassy** [reply]
Not supported yet @mrister. Is running on normal Fargate instances cost prohibitive? What sort of test volume do you have in mind?

**Reply by mrister** [reply]
It is not that prohibitive but we may have more testing done due to lower pricing. Perhaps up to 10k containers at some point.

**Comment by bernardobridge** [comment]
👋 The ability to use Spot instances is now available on [Artillery v2.0.5](https://github.com/artilleryio/artillery/releases/tag/artillery-2.0.5)

Note that to use this will require deleting the existing cluster, and letting Artillery recreate another. More information: https://www.artillery.io/docs/reference/cli/run-fargate#using---spot

c.c. @moufmouf @mrister
